NYPL/nypl-design-system
# DSD-1390: Vite update

Fixes JIRA ticket [DSD-1390](https://jira.nypl.org/browse/DSD-1390)

## This PR does the following:

This updates replaces TSDX with Vite for the build tool. This allows the Reservoir DS library to be easily maintained and allow individual npm packages to be updated for security patches and general bug fixes. At the same time, the build is now much faster. Specific updates and changes:

- Node 16 is the recommended version to use and it is set in the `.nvmrc` file.
- Github Actions are updated to use Node 16. The Typecheck workflow is removed since that's covered by the build script. Using Node 16 also allows the Vercel deployments to work without the the added `NODE_OPTIONS` flag in the npm build script.
- Vite is used to build the library and start a development build.
- Rollup plugins are removed in favor of Vite plugins.
- Gulp and the `resources.scss` file are removed. Other related `.scss` files were also removed since they are not used (in the library or in consuming applications).
- ESLint packages are updated.
- Jest is updated along with its config file. `jest-environment-jsom` and `ts-jest` are used now. Snapshot test output files have _slightly_ different rendered HTML syntax.
- The build is _still_ outputting CommonJS and ES module dist files, but the final name is slightly different. This does not affect how the DS is imported into any application.
- Storybook is updated to v6.5.16, which is a minor update from v6.5.13.
- Due to moving to ESM as the default for the repo, types are exported separated from components in the main `/src/index.ts` file. This doesn't affect the final build but just how components, interfaces, and types are exported.

## How to test

Make sure you have node 16 locally.

- Checkout the branch
- Run `nvm use`
- Delete `node_modules` and `package-lock.json`
- Run `npm i`
- Run all the npm scripts in `package.json`, such as `npm start`, `npm run build`, `npm test`, etc. They should all run and tests should pass.
  - When running `npm run build-storybook:v1`, make sure to then delete the `reservoir` folder as that should not be committed but also does not belong in the `.gitignore` file.
- Check the `dist` folder and make sure there are `design-system-react-components.cjs`, `design-system-react-components.js`, and `styles.css` files. Verify that there are type definitions in the `dist/src` directory.
- Verify the local storybook instance runs in the browser with `npm run storybook`.

## How has this been tested?

Locally.

If approved, we can merge this into the feature branch and create a release candidate. That npm version can be used to test in Turbine and other applications.

## Accessibility concerns or updates



- N/A, no functional changes to components are made.

### Checklist:




- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.

### Front End Review:





- [ ] Review the Vercel preview deployment once it is ready.
